configuration.md

```diff
@@ -0,0 +1,63 @@
+# Configuration
+You can configure the software using two different ways: Using a configuration file or
+environment variables. 
+
+## Config keys
+Both the configuration file and the environment variables share the same keys, listed below.
+
+|name	|type	|description	|required	|
+ --- | --- | --- | ---
+|REDIS_HOST	|String	|Host the Redis server runs at	|yes	|
+|REDIS_PORT	|Int	|Port for the Redis server	|yes	|
+|REDIS_PWD	|String	|Password for the Redis server, blank if none	|no	|
+|BUGZILLA_USER	|String	|Your openSUSE infrastructure username|yes	|
+|BUGZILLA_PASSWORD	|String	|Your openSUSE infrastructure password	|yes|
+|NO_LOAD	|String	|List of modules that should not be loaded. See below for details.	|yes, but can be empty|
+|TWITTER_CONSUMER_KEY	|String				||yes	|
+|TWITTER_CONSUMER_SECRET|String				||yes	|
+|TWITTER_ACCESS_TOKEN	|String 			||yes	|
+|TWITTER_ACCESS_TOKEN_SECRET	|String			||yes 	|
+
+## File based configuration
+Copy the file `config_sample.ini` to `config.ini`. It looks like this:
+
+```ini
+# Defrag sample configuration file
+# Copy to config.ini and replace the values!
+
+[REDIS]
+# Host the Redis server runs at
+REDIS_HOST = localhost
+
+# Port for the Redis server
+REDIS_PORT = 16052
+
+# Password for the Redis server, blank if none
+REDIS_PWD = password
+
+[BUGZILLA]
+# Your openSUSE infrastructure username
+BUGZILLA_USER = username
+
+# Your openSUSE infrastructure password
+BUGZILLA_PASSWORD = password
+
+[MODULES]
+# List of modules that should not be loaded
+# e.g. NO_LOAD = ["bugs", "twitter"]
+NO_LOAD = []
```

Comment:
Can you add the logic to handle this so I can see what you have mind?

Reply:
It's already there, in defrag/modules/__init__.py

Reply:
In defrag/init.py there is one _part_ of the implementation. I don't know how `__all__` is consumed so I don't know how the business logic is affected by the `__list_all_modules__` function...

Reply:
Another thing to bear in mind is that whoever edits the `.ini` file should not have to know anything about the business logic and dependences of the modules. But this requirement cannot be satisfied now and I am not sure that this requirement can be satisfied in the near future, simply because the `modules` directory contains things that don't depend on any modules and things that depend on some module(s).

I think it would be wise to first clean up the "dependency graph" before moving forward with the "NO LOAD" idea.

This means that perhaps it would be better to implement the "NO LOAD" idea on a different PR?

Reply:
I agree that this is not really clean, but not intended to be used by the end user anyway. I just want to have `NO_LOAD` so that I can test module without the others, e.g. when one module is broken but my branch is for working on another module, I can exclude the broken module.

Reply:
Okay, but then I am not sure about the use case. 
- all direct children of `modules` do not run any computation of their and have no state except a few string or dict assignations (which no computation run inbetween)
- all children of `modules` (including descendants) have their computation entirely driven by `ServiceManager.register`

What this means is that is that if you want to avoid computations/state, we can sort this out cleanly and simply in `__main__`, in particular in the `on_event("startup") handler.

As for ___programming mistakes___, I think it is not acceptable to ever have any one of these on `dev`. They should be catched upstream of `dev`, i.e. in PR-ed branches.

Reply:
Sure, but sometimes a module won't import even if no programming mistakes occur - Take the Twitter module for example: It won't work for me because I do not have the API keys etc, so it doesn't even need to get loaded. Also, using the service manager for disabling the module is even less clean of an implementation, as the endpoints are still registered to `app`. They just won't work, but are still there. 
What do you think of a compromise: I remove the `NO_LOAD` key from the documentation and the sample config, so that a user will never see it. But I can use it for making my programming life easier.
 
`__all__` is currently never consumed, I could as well remove it. The only thing consumed is `ALL_MODULES`, which is imported in `__main__`, get looped over while all the modules in there get imported.

Reply:
Thank you for being open minded about compromises :) 

I think removing the docs is a step in the right direction, but I am still not sure what benefit this achieves for you and for the code base. Your idea of making loading optional is a Good Idea, and I think this is something we should give more thoughts about. I don't have a beautiful solution right now, but I would be unhappy with a solution introducing differences between your code and `dev`, because debugging will get only harder and the purpose of git will be a little bit undermined.

Also where is the problem is you don't have API keys? I don't think we have any module that tries to do an action unless it exposes a function that is called. But in what circumstances would that happen if we avoid registering the module in the first place?

By the way there is a different use case that is more valid I think, namely lazy loading, aka load a module when and only when it's actually needed? No problem, a simple example of this approach can be found at https://levelup.gitconnected.com/python-trick-lazy-module-loading-df9b9dc111af (notice though that the use case for this is "faster loading because less stuff in memory" so this use case does not really apply to us since we have almost 0 state and 0 computation at importation time, and more importantly, this hides bugs; it could be good but I think later, when this project is more advanced).

So please let me know what you'd like me to change to make your life easier without introducing more deviation... 😸

Reply:
So can we keep the `NO_LOAD` feature for now (removed from the documentation) and replace it with something better in the future? My workflow is that I always disable or move away unneeded modules, and `NO_LOAD` makes that a lot easier. It might be that I just do so cause I'm used to it, but I really prefer to work this way.

Reply:
If that's so important for you and I don't get to think about it ever, I'd say why not.

Reply:
NO_LOAD is removed in 3a0cbc2. Archived in the `karatek/archive/noload` branch